### PR TITLE
JENKINS-35638 fix bug saving config.xml on a path with spaces in it

### PR DESCRIPTION
**Pull request by dalvizu**
[JENKINS-35638](https://issues.jenkins-ci.org/browse/JENKINS-35638)

### CODE REVIEW

File: core/src/main/java/hudson/model/ItemGroupMixIn.java [hunk 1]
```diff
@@ -271,7 +272,7 @@ public synchronized TopLevelItem createProjectFromXML(String name, InputStream x
         dir.mkdirs();
         boolean success = false;
         try {
-            XMLUtils.safeTransform((Source)new StreamSource(xml), new StreamResult(configXml));
+            XMLUtils.safeTransform((Source)new StreamSource(xml), new StreamResult(new FileOutputStream(configXml)));
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/ItemGroupMixIn.java]: 🐛 The stream won't be closed correctly in such case

**Comment by batmat** [comment]
For reference as the code seems weird at first, I find this explanation intesting http://stackoverflow.com/a/20436012/345845

**Comment by oleg-nenashev** [comment]
Yes. The code is reasonable. 1 fix is reqiured, right?

File: test/src/test/java/hudson/model/ItemGroupMixInTest.java [hunk 2]
```diff
@@ -207,4 +207,28 @@ public boolean isApplicable(Class jobType) {
         assertThat(Items.getConfigFile(foo).asString(), containsString("<description/>"));
     }
 
+    @Test
+    public void createProjectFromXMLWithSpaces() throws IOException {
+        final String xml = "<?xml version='1.0' encoding='UTF-8'?>\n" +
```
Comment (jtnord) [inline test/src/test/java/hudson/model/ItemGroupMixInTest.java]: this test is testing much more than the spaces.  The ENTITY use should be stripped here.

File: test/src/test/java/hudson/model/ItemGroupMixInTest.java [hunk 3]
```diff
@@ -207,4 +207,28 @@ public boolean isApplicable(Class jobType) {
...
+                "<!DOCTYPE project[\n" +
+                "  <!ENTITY foo SYSTEM \"file:///\">\n" +
+                "]>\n" +
+                "<project>\n" +
+                "  <actions/>\n" +
+                "  <description>&foo;</description>\n" +
```
Comment (jtnord) [inline test/src/test/java/hudson/model/ItemGroupMixInTest.java]: donlt use the entity - use a text description

File: test/src/test/java/hudson/model/ItemGroupMixInTest.java [hunk 4]
```diff
@@ -207,4 +207,28 @@ public boolean isApplicable(Class jobType) {
...
+                "  <canRoam>true</canRoam>\n" +
+                "  <triggers/>\n" +
+                "  <builders/>\n" +
+                "  <publishers/>\n" +
+                "  <buildWrappers/>\n" +
+                "</project>";
+
+        Item foo = r.jenkins.createProjectFromXML("foo bar", new ByteArrayInputStream(xml.getBytes()));
+        // if no exception then JAXP is swallowing these - so there should be no entity in the description.
+        assertThat(Items.getConfigFile(foo).asString(), containsString("<description/>"));
```
Comment (jtnord) [inline test/src/test/java/hudson/model/ItemGroupMixInTest.java]: test the description is correct after replacing the entity with a textual description

**Comment by jtnord** [comment]
I have lots of jobs with spaces in their names that work correctly....  What is special here?

**Comment by jtnord** [comment]
@batmat we where already passing in a `File` not a `String`.

This is a bug in a jaxp implementation somewhere - the System URI is being set to `file://some%20path`  and then the jaxp implementation is (I guess) doing the good old `URI.getFile()` which is doomed to fail.  So the thing here is to fix the broken jaxp implementation.  The default in JDK is ok - so something is providing garbage (and this is likely to cause a lot of other issues, so this is a sticky plaster fix, not a fix to the root cause AFAICT).  (See [JDK-5077403](https://bugs.openjdk.java.net/browse/JDK-5077403))

File: core/src/main/java/hudson/model/ItemGroupMixIn.java (hunk 1 above)
Comment (jtnord) [inline core/src/main/java/hudson/model/ItemGroupMixIn.java]: There is an underlying issue here.

> ` XMLUtils.safeTransform((Source)new StreamSource(xml), new StreamResult(configXml));`

This is not the correct way to construct a `StreamSource` from a String of XML content.

This should use the `StreamSource` constructor that takes either a `Reader` or an `InputStream`.

**Comment by jtnord** [comment]
So I could reproduce the issue with xalan and xerces on the classpath.
When they where both removed then the following code works as expected.

```java
import java.io.File;
import java.io.StringReader;

import javax.xml.transform.Source;
import javax.xml.transform.stream.StreamResult;
import javax.xml.transform.stream.StreamSource;

import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;

import jenkins.util.xml.XMLUtils;

import static org.junit.Assert.assertTrue;

public class Scrath {

	@Rule
	public TemporaryFolder tempFile = new TemporaryFolder();
	
	@Test
	public void someTest() throws Exception {
		File parent = tempFile.newFolder("With Spaces");
		assertTrue("parent should exist", parent.exists());
		File f = new File(parent, "config.xml");
		final String xml = "<?xml version='1.0' encoding='UTF-8'?>\n" +
		                   "<project>\n" +
		                   "  <actions/>\n" +
		                   "  <description>blah</description>\n" +
		                   "  <keepDependencies>false</keepDependencies>\n" +
		                   "  <properties/>\n" +
		                   "  <scm class=\"hudson.scm.NullSCM\"/>\n" +
		                   "  <canRoam>true</canRoam>\n" +
		                   "  <triggers/>\n" +
		                   "  <builders/>\n" +
		                   "  <publishers/>\n" +
		                   "  <buildWrappers/>\n" +
		                   "</project>";
		XMLUtils.safeTransform((Source)new StreamSource(new StringReader(xml)), new StreamResult(f));
	}
}
```

Thus the original code is (mostly apart from the construction of a `StreamSource`) correct.

For plugins unit tests at least these come in from htmlunit via jenkins-test-harness, what is pulling these in at runtime is a different question.

**Comment by dalvizu** [comment]
@jtnord thanks for review - will address shortly
re:
> I have lots of jobs with spaces in their names that work correctly.... What is special here?

Sorry for not including some more context: I was seeing this issue in a project I have to test job-dsl files. The test runs jenkins with @JenkinsRule and my pipeline files with spaces in the job names would return this error. Those same pipelines would run on actual Jenkins servers just fine.

**Comment by jtnord** [comment]
> Those same pipelines would run on actual Jenkins servers just fine.

Add the following to your POM to exclude the evil dependencies from htmlunit.

```xml
  </dependencies>
	<dependencies>
		<dependency>
			<groupId>org.jenkins-ci.main</groupId>
			<artifactId>jenkins-test-harness</artifactId>
			<scope>test</scope>
			<exclusions>
				<exclusion>
					<groupId>junit</groupId>
					<artifactId>junit-dep</artifactId>
				</exclusion>
				<exclusion>
					<groupId>xalan</groupId>
					<artifactId>xalan</artifactId>
				</exclusion>
				<exclusion>
					<groupId>xerces</groupId>
					<artifactId>xercesImpl</artifactId>
				</exclusion>
			</exclusions>
		</dependency>
	</dependencies>
```

Not sure if `htmlunit` really needs these for a modern JDK.  Regardless I could not find an issue in xalan or xerces for this issue, but the stock parsers in the JDK 8 (oracle 1.8.0_91 x64 on windows) work for me.

**Comment by dalvizu** [comment]
Excluding those dependencies gives me AbstractMethodError:

```
Expected no exception to be thrown, but got 'java.lang.AbstractMethodError'
	at spock.lang.Specification.noExceptionThrown(Specification.java:118)
	at com.pingone.JobScriptsSpec.test script #file.name(JobScriptsSpec.groovy:35)
Caused by: java.lang.AbstractMethodError: javax.xml.transform.sax.SAXTransformerFactory.setFeature(Ljava/lang/String;Z)V
	at jenkins.util.xml.XMLUtils.safeTransform(XMLUtils.java:46)
	at hudson.model.ItemGroupMixIn.createProjectFromXML(ItemGroupMixIn.java:279)
	at jenkins.model.Jenkins.createProjectFromXML(Jenkins.java:3109)
	at javaposse.jobdsl.plugin.JenkinsJobManagement.createNewItem(JenkinsJobManagement.java:529)
	at javaposse.jobdsl.plugin.JenkinsJobManagement.createOrUpdateConfig(JenkinsJobManagement.java:135)
	at javaposse.jobdsl.dsl.DslScriptLoader.extractGeneratedJobs(DslScriptLoader.java:150)
	at javaposse.jobdsl.dsl.DslScriptLoader.runDslEngineForParent(DslScriptLoader.java:88)
	at javaposse.jobdsl.dsl.DslScriptLoader.runDslEngine(DslScriptLoader.java:132)
	at javaposse.jobdsl.dsl.DslScriptLoader.runDslEngine(DslScriptLoader.java:128)
	at com.pingone.JobScriptsSpec.test script #file.name(JobScriptsSpec.groovy:32)
```

**Comment by andresrc** [comment]
@dalvizu that error seems to be that by excluding the dependency an even older version came into the classpath.

**Comment by oleg-nenashev** [comment]
@andresrc @jtnord From what I see this PR still needs some work, right?

**Comment by jtnord** [comment]
@oleg-nenashev it's not that the PR needs some work, it's that it seems to be due to an invalid setup.
Now the fact that this happens in test code only and the changes fix it you may think "well this is safe" however the test code failing then means you are not testing code in the same way that it would be tested when run in Jenkins which to me would be a concern.
Basically something is pulling in XML parsers[1] that are replacing the JDK parsers and that is causing the issue.  Killing all the inclusion of the XML parsers is what we should be doing.  It's 2016, there is no reason to not be using the parser that ships with the JDK :)

[1] that are either buggy but implement the new API, or are so old they do not implement the new API from the spec shipped with the JDK

**Comment by daniel-beck** [comment]
It appears this PR has been abandoned and may have no merit to it (previous comment).

**Comment by dalvizu** [comment]
I would still like to see this merged. @jtnord has objections, but I am not sure they can be addressed. 

Jenkins ships with XML parsers which are buggy, this works around that bug, and "Killing all the inclusion of the XML parsers" is a much larger change as to be inactionable. This solves a very real, very reproducible issue with jenkins test setup without any side-effects. I don't consider this to be abandoned, or without merit.

**Comment by daniel-beck** [comment]
@dalvizu Noted.

**Comment by daniel-beck** [comment]
@jtnord @oleg-nenashev Is this a symptom of [JENKINS-30685](https://issues.jenkins-ci.org/browse/JENKINS-30685)?

**Comment by jtnord** [comment]
@daniel-beck I think that this is that the plugin is bringing in garbage XML parsers rather than Jenkins sucking them in - so is different from JENKINS-30685

**Comment by oleg-nenashev** [comment]
@daniel-beck @jtnord I am fine with this change since it works around the case. Classloading in the core is definitely a problem, but it should be rather solved in Jenkins 3 instead of this PR

**Review by oleg-nenashev — APPROVED** [review]

**Comment by daniel-beck** [comment]
Plenty of changes in this weekly already, so let's hold for next week?

**Review by jtnord — COMMENTED** [review]
@oleg-nenashev  - I think you need to go back through the history - especially my comment from Jun 29, 2016

There are no steps to reproduce this issue in the associated JIRA.

Both myself and @andresrc of of the opinion that this is an error caused by the users test job (or perhaps jobdsl) bringing in buggy libraries and not Jenkins itself which does the right thing - as can be validated if you look at your employers CI system which has many jobs with spaces in the names.

There are no steps to reproduce the issue included in the bug report which can be used verify that indeed it was the test setup (JenkinsRule does different classloading)  that was the issue - and brining in old XML parsers is likely to just cause other things to break as well - so this is just masking the issue of a buggy parser and making the code more obscure.

Have you tried the test without the fix - does it fail for you?  If not then this demonstrates that Jenkins itself is not the issue.

**Comment by dalvizu** [comment]
@jtnord - it was `jenkins-test-harness` which was bringing in these dependencies (xalan, xerces). When I excluded them from `jenkins-test-harness` per your instructions, I would receive `AbstractMethodError`. So I assume the ones from Jenkins are being used.

If I were to provide an end-to-end example project, would that help address your concerns?

**Comment by jtnord** [comment]
@dalvizu example project would be great

**Comment by jgangemi** [comment]
i ran into this issue similar to @dalvizu testing some job dsl scripts.

i threw together a sample project that exhibits the behavior. you'll need to build it using maven instead of gradle.

https://github.com/jgangemi/jenkins-35638-issue

**Comment by daniel-beck** [comment]
@jtnord See previous comment.

**Comment by jtnord** [comment]
the fix I outlined above (excluding xerces and xalan) when added to the project shows that the build passes as expected.  Ergo suggesting closing as originaly discussed.  (See jgangemi/jenkins-35638-issue#1)

I suggest that maybe the test harness should have these XML libraries excluded - as there should be almost no need for anything to use these in a JDK8 world... (but maybe an old dep is using internal apis rather than the javax api)

**Comment by jtnord** [comment]
@jgangemi @dalvizu See my latest comment - not an issue in Jenkins, an issue with bad / buggy parsers being pulled in (by jenkins-test-harness).

/cc @daniel-beck

**Comment by daniel-beck** [comment]
@jtnord Do we need to implement the exclusion e.g. in the test harness? Or is this different and unrelated to anything we publish?

**Comment by jtnord** [comment]
@daniel-beck I could have sworn I opened a JIRA about it.
RIght now I do not know what xerces etc are needed for in the test harness.  There may be some requirement of an old lib that uses the xerces API rather than the jax one such that just excluding them may break some other test harness functionality (htmlunit or something).  So whilst excluding them should be the goal, (IMO we should never have any XML parsers in the the build and should rely on core JDK ones) we need to be careful and possibly update other things along the way to make sure we can remove them safely for all uses.

**Comment by oleg-nenashev** [comment]
As there is no comments from @jgangemi regarding the analysis by @jtnord, I am closing this pull request

**Pull request closed without merging**